### DISCUSSION TITLE
AppShell fixed property doesn't work if using custom Navbar component

### DISCUSSION THREAD

**Discussion by alukach** [mantinedev/mantine#399] [Q&A] [answered]
## Issue

The [`AppShell` docs](https://mantine.dev/core/app-shell/) demonstrate using a `Navbar` component in the `navbar` prop and using the `fixed` prop to allow the application to be offset from the navbar.  This works well.

<details>

<summary>As demonstrated in docs: Code + Demo Video</summary>

```ts
import { AppShell, Navbar, Title } from '@mantine/core';
import React from 'react';

export function Layout({ children }: React.PropsWithChildren<{}>) {
  return <AppShell
    fixed
    navbar={
      <Navbar
        width={{ base: 250, breakpoints: { sm: '100%', md: 275 } }}
        padding="xs"
      >
        <Navbar.Section>
          <Title>My App</Title>
        </Navbar.Section>

        <Navbar.Section grow mt="lg">
          Links...
        </Navbar.Section>

        <Navbar.Section>
          User...
        </Navbar.Section>
      </Navbar>
    }
  >
    {children}
  </AppShell >
}
```

https://user-images.githubusercontent.com/897290/140998080-ea7bfdc0-9af8-4700-9c9b-5f9ba3a68eac.mov
</details>

However, when a users `Navbar` starts becoming somewhat complicated, they may want to break it out into a custom component for the sake of encapsulation (e.g. `MyNavbar`).  However once this is done, **the `fixed` prop on the `AppShell` no longer correctly fixes the `Navbar`.**

<details>

<summary>Code + Demo Video</summary>

```ts
import { AppShell, Navbar, Title } from '@mantine/core';
import React from 'react';

// NOTE: Broke out navbar into a custom component
const MyNavbar: React.FC = () => (
  <Navbar
    width={{ base: 250, breakpoints: { sm: '100%', md: 275 } }}
    padding="xs"
  >
    <Navbar.Section>
      <Title>My App</Title>
    </Navbar.Section>

    <Navbar.Section grow mt="lg">
      Links...
    </Navbar.Section>

    <Navbar.Section>
      User...
    </Navbar.Section>
  </Navbar>
)

export function Layout({ children }: React.PropsWithChildren<{}>) {
  return <AppShell
    fixed
    navbar={<MyNavbar />}
  >
    {children}
  </AppShell >
}
```

https://user-images.githubusercontent.com/897290/140998258-42144632-ab21-4848-a472-323cf6121b87.mov
</details>

A user may attempt to overcome this by manually placing the `fixed` property on their custom component. However, **placing a `fixed` property on the `Navbar` does not correctly offset the application.**

<details>

<summary>Code + Demo Video</summary>

```ts
import { AppShell, Navbar, Title } from '@mantine/core';
import React from 'react';

const MyNavbar: React.FC = () => (
  <Navbar
    fixed // Note: Manually fixing navbar in custom component
    width={{ base: 250, breakpoints: { sm: '100%', md: 275 } }}
    padding="xs"
  >
    <Navbar.Section>
      <Title>My App</Title>
    </Navbar.Section>

    <Navbar.Section grow mt="lg">
      Links...
    </Navbar.Section>

    <Navbar.Section>
      User...
    </Navbar.Section>
  </Navbar>
)

export function Layout({ children }: React.PropsWithChildren<{}>) {
  return <AppShell
    fixed
    navbar={<MyNavbar />}
  >
    {children}
  </AppShell >
}
```
  

https://user-images.githubusercontent.com/897290/140998473-b3b3830f-87ad-41a6-87d5-fef91feede24.mov
</details>

## Question

Is it possible to properly set up a custom Navbar component so that the `AppShell` component's `fixed` property properly offsets the application from the navbar?

**Comment by rtivital** [comment, marked as answer]
To use Navbar and Header components with your wrapper you can do two things:

1. Add Navbar props support in your component:
```tsx
import { AppShell, Navbar, NavbarProps } from '@mantine/core';

function MyNavbar(props: NavbarProps) {
  return <Navbar {...props}>Your component</Navbar>
}

function MyAppShell() {
  return <AppShell fixed navbar={<MyNavbar width={{ base: 250, breakpoints: { sm: '100%', md: 275 } }} padding="xs" />} />
}
```

2. Use your component without Navbar:

```tsx
import { AppShell, Navbar, NavbarProps } from '@mantine/core';

function MyNavbar() {
  return <div>Your component</div>
}

function MyAppShell() {
  return <AppShell fixed navbar={<Navbar width={{ base: 250, breakpoints: { sm: '100%', md: 275 } }} padding="xs"><MyNavbar /></Navbar>} />
}
```

AppShell requires direct props be present on child component, it will not work with component that does not have any props.

**Reply by alukach** [reply]
This totally makes sense.  Thanks!

(Loving this library, btw)

**Reply by rtivital** [reply]
Thanks!

**Reply by mumingpo** [reply]
Is this no longer relevant since 4.0.3?

[4.0.3 patch notes](https://github.com/mantinedev/mantine/releases/tag/4.0.3)
